### PR TITLE
fix: harden revision sync release gates

### PR DESCRIPTION
**Pull request by sachaservan**
## Summary
- persist account-scoped remote-delete tombstones before revision checkpoints can advance, reject stale saves/uploads after relaunch, and retry local cleanup safely
- centralize logical edit clock stamping for cloud chat saves and stamp frozen upload payloads for the expected CAS version
- finalize successful uploads with authoritative clock metadata while preserving newer edits that race an upload
- add pure policy coverage for upload clock stamping and finalization

## Validation
- inspected the complete diff against `origin/main`
- ran `git diff --check origin/main...HEAD`
- did not run builds, compilation, or tests per `AGENTS.md`
- manual Xcode testing is required

---
## Summary by cubic
Hardens revision sync gates and finalization so remote deletes persist across restarts, clocks remain trusted and monotonic, and version bumps are bounded. Previously, deletes could be lost and uploads could finalize with stale clocks or overflowed versions; now we persist encrypted tombstones, gate writes/uploads, serialize clock allocation, reconcile final clocks, and validate versions (including on metadata-only saves).

- Persist encrypted remote-delete tombstones per account; block saves/uploads when tombstoned or upgrade‑gated via `CloudUploadGate`/`ensureChatUploadIsAllowed`; defer local cleanup while streaming; clear tombstones only after verified absence or authoritative remote apply; restore and retry across relaunches with `requiresRemoteDeleteRecovery`/`completeRemoteDeleteRecovery`.
- Enforce clock/version policy: `EditClockAllocator` serializes ticks and throws on exhaustion; `ChatEditClockPolicy` stamps upload payloads, validates finalization via `uploadedClock`, and preserves edits that race uploads; local saves validate sync versions and keep trusted clocks for metadata‑only changes; sync‑version increments use bounded `nextSyncVersion`.
- Strengthen reconciliation: treat known tombstones as must‑pull, allow authoritative upserts to replace local deletes when tombstoned, and only then clear tombstones.

**Review/migration**
- Handle `EditClockStore.nextClock` throwing (updated in `ProfileManager` and `ChatRecoverySync`).
- Update upload paths: stamp via `stampClockVersionForUpload`, pass `uploadedClock` to `finalizeUploadIfFresh`, and gate via `CloudUploadGate` and `ensureChatUploadIsAllowed`.
- Treat `EncryptedFileStorage.saveChat` throwing `SaveError.remotelyDeleted` as a cancel/no‑op; handle `SaveError.invalidSyncVersion` as conflict/invalid revision.
- Guard version increments with `ChatEditClockPolicy.nextSyncVersion` and handle `nil` as conflict/invalid revision.

<sup>Written for commit f3048e9502119bc9db0eb0365f2314e4392c51b7. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/tinfoilsh/tinfoil-ios/pull/260?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: TinfoilChat/Services/EditClock.swift [hunk 1]
```diff
@@ -18,6 +18,48 @@ struct EditClock: Codable, Equatable {
...
+enum ChatEditClockPolicy {
+    static func uploadState(
+        clock: Int?,
+        writer: String?,
+        currentSyncVersion: Int
+    ) -> ChatClockState {
+        ChatClockState(
+            clock: clock,
+            writer: writer,
+            clockVersion: clock == nil || writer == nil ? nil : currentSyncVersion + 1
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EditClock.swift]: Fixed in 1d0d490: upload/finalize only carry clocks whose source clockVersion is already trusted; stale or missing versions remain untrusted.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 2]
```diff
@@ -577,11 +577,18 @@ class CloudSyncService: ObservableObject {
...
+                          userId: userId
+                      ) else {
                     throw CancellationError()
                 }
+                let uploadChat = stampClockVersionForUpload(
+                    StoredChat(from: chat, syncVersion: chat.syncVersion)
+                )
                 let result = try await cloudStorage.uploadChat(
-                    StoredChat(from: chat, syncVersion: chat.syncVersion),
+                    uploadChat,
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:604]: Fixed in 1d0d490: all upload entrypoints now use the shared CloudUploadGate, including final pre-upload checks.

File: TinfoilChat/Services/EncryptedFileStorage.swift [hunk 3]
```diff
@@ -425,6 +515,9 @@ actor EncryptedFileStorage {
     }
 
     private func performSaveChat(_ chat: Chat, userId: String) async throws {
+        guard try await !hasRemoteDeleteTombstone(chatId: chat.id, userId: userId) else {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EncryptedFileStorage.swift]: Outdated: current applyRemoteChatDelete persists and marks the tombstone before both streaming and non-streaming cleanup.

File: TinfoilChat/Services/EncryptedFileStorage.swift [hunk 4]
```diff
@@ -302,7 +312,56 @@ actor EncryptedFileStorage {
     func saveChat(_ chat: Chat, userId: String) async throws {
         await acquireWriteLock()
         defer { releaseWriteLock() }
-        try await performSaveChat(chat, userId: userId)
+        let prepared = try await prepareLocalChatForSave(chat, userId: userId)
+        try await performSaveChat(prepared, userId: userId)
+    }
+
+    private func prepareLocalChatForSave(_ chat: Chat, userId: String) async throws -> Chat {
+        guard subdirectory == nil, chat.locallyModified else { return chat }
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EncryptedFileStorage.swift:332]: Rejected: Swift guard continues only when both conditions are true, so cloud + locallyModified is exactly the path that enters stamping.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 5]
```diff
@@ -593,6 +600,7 @@ class CloudSyncService: ObservableObject {
                     userId: userId,
                     expectedUpdatedAt: chat.updatedAt,
                     syncVersion: newVersion,
+                    uploadedClock: clockState(uploadChat),
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:619]: Fixed in 1d0d490: reportChatSynced is now preceded by generation and current-user checks after finalization.

File: TinfoilChat/Services/EditClock.swift [hunk 6]
```diff
@@ -18,6 +18,48 @@ struct EditClock: Codable, Equatable {
...
+    let clockVersion: Int?
+}
+
+enum ChatEditClockPolicy {
+    static func uploadState(
+        clock: Int?,
+        writer: String?,
+        currentSyncVersion: Int
+    ) -> ChatClockState {
+        ChatClockState(
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EditClock.swift]: Rejected as code-smell-only: the three branches encode distinct fresh/racing version semantics and are clearer inline.

File: TinfoilChat/Services/EncryptedFileStorage.swift [hunk 7]
```diff
@@ -302,7 +312,56 @@ actor EncryptedFileStorage {
...
+        let encPath = try chatFilePath(chatId: chat.id, userId: userId, isCorrupted: false)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EncryptedFileStorage.swift:333]: Rejected as non-blocking extraction churn; the lock-scoped call sites intentionally make their file and sidecar sequencing explicit.

File: TinfoilChat/Services/ChatRecoverySync.swift [hunk 8]
```diff
@@ -140,13 +140,18 @@ actor ChatRecoverySync {
...
                 candidate.syncVersion = remote.syncVersion
-                stampEdit(&candidate, observedRemote: remoteChat)
+                try stampEdit(&candidate, observedRemote: remoteChat)
                 candidate.locallyModified = true
 
                 guard await Clerk.shared.user?.id == userId else {
                     throw ChatRecoverySyncError.chatMissing
                 }
                 try Task.checkCancellation()
+                let isRemoteDeleted = try await EncryptedFileStorage.cloud.hasRemoteDeleteTombstone(
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ChatRecoverySync.swift:150]: Fixed in 1d0d490: cancellation and current-user identity are rechecked after the tombstone/gate awaits and immediately before upload.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 9]
```diff
@@ -1663,13 +1698,38 @@ class CloudSyncService: ObservableObject {
...
+                )
                 deferredRemoteDeletes.removeValue(forKey: chatId)
             } catch {
                 // Still failing; keep the entry for the next cycle.
             }
         }
     }
 
+    private func restoreDurableRemoteDeletes(userId: String, generation: Int) async throws {
+        let tombstones = try await EncryptedFileStorage.cloud.loadRemoteDeleteTombstones(
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:1745]: Outdated: malformed metadata is quarantined, account uploads remain fenced, and a successful normal snapshot clears recovery. A transient uninitialized key still intentionally retries without discarding metadata.

File: TinfoilChat/Services/ChatRecoverySync.swift [hunk 10]
```diff
@@ -140,13 +140,18 @@ actor ChatRecoverySync {
...
+                    chatId: chatId,
+                    userId: userId
+                )
+                guard !isRemoteDeleted else { throw CancellationError() }
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ChatRecoverySync.swift]: Fixed in 1d0d490: a tombstoned recovery chat now throws ChatRecoverySyncError.chatMissing.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 11]
```diff
@@ -952,10 +969,10 @@ class CloudSyncService: ObservableObject {
             }
 
             let localClock = localChat.flatMap {
-                trustedClock(
-                    clock: $0.clock, writer: $0.writer,
-                    clockVersion: $0.clockVersion, syncVersion: $0.syncVersion
-                )
+                let expectedClockVersion = $0.syncVersion + ($0.locallyModified ? 1 : 0)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift]: Fixed in 1d0d490: local conflict arbitration now calls trustedChatClock directly.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 12]
```diff
@@ -577,11 +577,19 @@ class CloudSyncService: ObservableObject {
...
+                      try await !EncryptedFileStorage.cloud.hasRemoteDeleteTombstone(
+                          chatId: chat.id,
+                          userId: userId
+                      ) else {
                     throw CancellationError()
                 }
+                let uploadChat = stampClockVersionForUpload(
+                    StoredChat(from: chat, syncVersion: chat.syncVersion)
+                )
+                try await ensureChatUploadIsAllowed(chatId: chat.id, userId: userId)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:602]: Fixed in 1d0d490: the redundant inline delete checks were removed; the shared final eligibility check remains immediately before upload.

File: TinfoilChat/Services/EncryptedFileStorage.swift [hunk 13]
```diff
@@ -584,26 +732,47 @@ actor EncryptedFileStorage {
...
+        updatedEntries.removeAll { $0.id == chatId }
+        if updatedEntries != entries {
+            try await saveIndex(updatedEntries, userId: userId)
+        }
+        contentRepairIds[userId]?.remove(chatId)
+
+        let contentIsAbsent = try !hasChatContentFile(chatId: chatId, userId: userId)
+        let sidecarPath = try syncMetadataPath(chatId: chatId, userId: userId)
+        let legacySidecarPath = try legacySyncMetadataPath(chatId: chatId, userId: userId)
+        return RemoteDeleteTombstonePolicy.canClearAfterLocalCleanup(
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EncryptedFileStorage.swift:780]: Rejected: the post-removal stat is the authoritative proof that orphan content and sidecars are absent before clearing a durable delete-wins tombstone; correctness justifies the redundant check.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 14]
```diff
@@ -1769,14 +1841,16 @@ class CloudSyncService: ObservableObject {
                 )
             },
             recentLimit: Constants.Pagination.chatsPerPage,
-            missingContentIds: missingContentIds
+            missingContentIds: missingContentIds,
+            knownTombstoneIds: remoteDeleteIds
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:1868]: Rejected: account-wide recovery intentionally runs normal snapshot planning. Remote-absent known-server rows are removed, while remotely present dirty rows remain CAS-protected and must not be overwritten.

File: TinfoilChat/Services/EditClock.swift [hunk 15]
```diff
@@ -13,69 +13,159 @@ import Foundation
...
+        return deviceIdUnlocked()
+    }
 
-    /// Stable id for this installation. Generated once and persisted.
-    static func deviceId() -> String {
-        if let existing = defaults.string(forKey: deviceIdKey), !existing.isEmpty {
+    func observe(_ remoteV: Int?) {
+        lock.lock()
+        defer { lock.unlock() }
+        guard let remoteV, remoteV > 0, remoteV < EditClockStore.maxCounter else { return }
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EditClock.swift:38]: Rejected as astronomical adversarial defense-in-depth. The current policy rejects exhausted remote input and never wraps/reuses a local tuple; introducing an arbitrary lower high-water mark would reject valid cross-platform clocks.

File: TinfoilChat/Services/ProfileManager.swift [hunk 16]
```diff
@@ -678,7 +681,13 @@ class ProfileManager: ObservableObject {
         }
 
         guard generation == accountGeneration else { return }
-        var profile = prepareLocalProfileForSync()
+        var profile: ProfileData
+        do {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ProfileManager.swift:685]: Rejected: try! would violate the explicit no-wrap/no-reuse invariant. Genuine persisted exhaustion must refuse sync safely, even though it is practically unreachable.

File: TinfoilChatTests/RevisionSyncTests.swift [hunk 17]
```diff
@@ -534,6 +596,116 @@ struct RevisionSyncTests {
...
+    }
+
+    @Test func editClockRefusesCounterExhaustion() {
+        #expect(throws: EditClockStore.ClockError.self) {
+            _ = try EditClockStore.incrementedCounter(after: EditClockStore.maxCounter)
+        }
+    }
+
+    @Test func editClockAllocatorSerializesConcurrentTicks() {
+        let suite = "edit-clock-concurrency-\(UUID().uuidString)"
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChatTests/RevisionSyncTests.swift:696]: Rejected as test-only helper churn; the isolated suite setup is short and each defer keeps cleanup adjacent to ownership.

File: TinfoilChat/Services/ProfileManager.swift [hunk 18]
```diff
@@ -609,9 +609,12 @@ class ProfileManager: ObservableObject {
             if let cloudProfile = try await profileSync.fetchProfile() {
                 guard generation == accountGeneration else { return false }
                 let cloudVersion = cloudProfile.version ?? 0
-                let localProfile = hasPendingLocalProfileChanges
-                    ? prepareLocalProfileForSync()
-                    : createProfileData()
+                let localProfile: ProfileData
+                if hasPendingLocalProfileChanges {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ProfileManager.swift:613]: Rejected: this can fail only at genuine locally persisted clock exhaustion; merging dirty local profile state without a fresh clock is unsafe, so the pull must refuse rather than silently apply.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 19]
```diff
@@ -1528,54 +1563,53 @@ class CloudSyncService: ObservableObject {
...
-                userId: userId,
-                generation: generation,
-                preserveNeverSynced: preserveNeverSynced,
-                callbackRegistered: false
-            )
+            userId: userId,
+            preserveNeverSynced: preserveNeverSynced
         )
-        return false
+        guard generation == accountGeneration else { throw CancellationError() }
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:1612]: Fixed in 1d0d490: intentional local-only/never-synced preservation now clears the durable tombstone, deferred entry, and in-memory delete marker. Known-server rows still require deletion.

File: TinfoilChat/Services/EditClock.swift [hunk 20]
```diff
@@ -13,69 +13,159 @@ import Foundation
...
+        }
+    }
+
+    func nextClock(observedMax: Int? = nil) throws -> EditClock {
+        lock.lock()
+        defer { lock.unlock() }
+        let validObservedMax = observedMax.flatMap {
+            $0 > 0 && $0 < EditClockStore.maxCounter ? $0 : nil
+        } ?? 0
+        let base = min(
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EditClock.swift:50]: Rejected as non-functional cleanup; retaining the explicit ceiling clamp documents and defends the allocator bound without changing behavior.

File: TinfoilChat/Services/EncryptedFileStorage.swift [hunk 21]
```diff
@@ -873,6 +1159,15 @@ actor EncryptedFileStorage {
         return removedAll
     }
 
+    private func removeSyncSidecarsStrict(chatId: String, userId: String) throws {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EncryptedFileStorage.swift:1172]: Rejected as code-smell-only. Strict remote-delete cleanup and best-effort prune intentionally have different failure semantics; keeping them explicit avoids weakening tombstone clearing.

File: TinfoilChat/Services/EditClock.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EditClock.swift:38]: Rejected: remote values at the JS ceiling are deliberately treated as untrusted because accepting one would exhaust local allocation. A genuinely local persisted ceiling still refuses reuse.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 22]
```diff
@@ -1657,19 +1691,65 @@ class CloudSyncService: ObservableObject {
...
+        chatId: String,
+        pending: DeferredRemoteDelete
+    ) async throws -> Bool {
+        let confirmedAbsent = try await EncryptedFileStorage.cloud.removeChatForConfirmedRemoteDelete(
+            chatId: chatId,
+            userId: pending.userId,
+            preserveNeverSynced: pending.preserveNeverSynced
+        )
+        guard pending.generation == accountGeneration else { throw CancellationError() }
+        guard confirmedAbsent else {
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:1722]: Rejected after tracing current storage semantics: false is returned only for intentional local-only/never-synced preservation. Content or sidecar removal failures throw, so the tombstone and deferral remain for retry.

File: TinfoilChat/Services/ChatRecoverySync.swift [hunk 23]
```diff
@@ -140,13 +140,23 @@ actor ChatRecoverySync {
...
                 guard await Clerk.shared.user?.id == userId else {
                     throw ChatRecoverySyncError.chatMissing
                 }
                 try Task.checkCancellation()
+                let isRemoteDeleted = try await EncryptedFileStorage.cloud.hasRemoteDeleteTombstone(
+                    chatId: chatId,
+                    userId: userId
+                )
+                guard !isRemoteDeleted else { throw ChatRecoverySyncError.chatMissing }
+                _ = try await CloudUploadGate.allowsWrite(required: true)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/ChatRecoverySync.swift:155]: Fixed in a426b93: recovery now rechecks the durable tombstone after the gate and identity awaits immediately before upload.

File: TinfoilChat/Services/EditClock.swift [hunk 24]
```diff
@@ -13,69 +13,192 @@ import Foundation
...
+        current: ChatClockState,
+        uploaded: ChatClockState
+    ) -> Bool {
+        current.clock == uploaded.clock && current.writer == uploaded.writer
+    }
+
+    static func uploadState(
+        clock: Int?,
+        writer: String?,
+        sourceClockVersion: Int?,
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EditClock.swift:119]: Rejected as refactor-only: all call sites use argument labels, and the current names distinguish source state from current persisted state without a behavioral ambiguity.

File: TinfoilChat/Services/EditClock.swift [hunk 25]
```diff
@@ -13,69 +13,192 @@ import Foundation
...
+
+enum ChatEditClockPolicy {
+    static func isTrusted(
+        _ state: ChatClockState,
+        syncVersion: Int,
+        locallyModified: Bool
+    ) -> Bool {
+        state.clock != nil
+            && state.writer?.isEmpty == false
+            && state.clockVersion == syncVersion + (locallyModified ? 1 : 0)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EditClock.swift]: Fixed in a426b93: sync-version increments are now checked, malformed boundary values remain untrusted/refused, and source boundary tests were added.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 26]
```diff
@@ -1657,19 +1691,65 @@ class CloudSyncService: ObservableObject {
...
+            try await EncryptedFileStorage.cloud.clearRemoteDeleteTombstone(
+                chatId: chatId,
+                userId: pending.userId
+            )
+            deferredRemoteDeletes.removeValue(forKey: chatId)
+            deletedChatsTracker.removeFromDeleted(chatId)
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:1729]: Fixed in a426b93: both intentional-preservation and confirmed-removal branches recheck account generation after durable tombstone clearing before mutating in-memory trackers.

File: TinfoilChat/Services/EncryptedFileStorage.swift [hunk 27]
```diff
@@ -302,7 +324,63 @@ actor EncryptedFileStorage {
...
+            && chat.clock.map { $0 > (existing?.clock ?? 0) } == true
+            && chat.writer?.isEmpty == false
+        if !carriesNewClock {
+            let clock = try EditClockStore.nextClock(
+                observedMax: max(existing?.clock ?? 0, chat.clock ?? 0)
+            )
+            prepared.clock = clock.v
+            prepared.writer = clock.w
+        }
+        guard let preparedClockVersion = ChatEditClockPolicy.nextSyncVersion(
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/EncryptedFileStorage.swift]: Fixed in f3048e9: sync-version validation now runs before the metadata-only early return, so malformed boundaries cannot be persisted through that path.

File: TinfoilChat/Services/CloudSyncService.swift [hunk 28]
```diff
@@ -577,22 +595,28 @@ class CloudSyncService: ObservableObject {
...
                 let result = try await cloudStorage.uploadChat(
-                    StoredChat(from: chat, syncVersion: chat.syncVersion),
+                    uploadChat,
                     idempotencyKey: newSyncEnclaveIdempotencyKey()
                 )
                 guard generation == accountGeneration else {
                     throw SyncEnclaveError(message: "account changed during cloud backup")
                 }
-                let newVersion = result.syncVersion ?? chat.syncVersion + 1
+                guard let newVersion = result.syncVersion
```
[the remark this answers is not in the document]
Comment (sachaservan) [reply TinfoilChat/Services/CloudSyncService.swift:610]: Rejected: this is two short guards at distinct upload entry points, and keeping each path locally explicit is clearer than a helper coupling CloudStorageService.UploadChatResult to generic version arithmetic. Both already use the single ChatEditClockPolicy.nextSyncVersion policy, so behavior cannot drift independently.
